core/src/nucleus/ribosome.rs

```diff
@@ -38,35 +36,51 @@ enum HcApiFuncIndex {
 }
 
 /// HcApiFuncIndex::PRINT function code
+/// args: [0] encoded MemoryAllocation as u32
+/// Expecting a string as complex input argument
+/// Returns an HcApiReturnCode as I32
 fn invoke_print(runtime: &mut Runtime, args: &RuntimeArgs) -> Result<Option<RuntimeValue>, Trap> {
-    let arg: u32 = args.nth(0);
-    runtime.print_output.push(arg);
+    assert!(args.len() == 1);
+
+    // Read complex argument serialized in memory
+    let encoded_allocation: u32 = args.nth(0);
```

Comment:
this whole section, asserting the args, getting the allocation, reading it as utf8, seems like a potentially nice util function, but i'm happy to do that in my branch :)

Reply:
+1